src/backend/distributed/commands/table.c

```diff
@@ -819,23 +822,36 @@ PreprocessAlterTableStmt(Node *node, const char *alterTableCommand,
 		else if (alterTableType == AT_AttachPartition)
 		{
 			PartitionCmd *partitionCommand = (PartitionCmd *) command->def;
-
-			/*
-			 * We only support ALTER TABLE ATTACH PARTITION, if it is only subcommand of
-			 * ALTER TABLE. It was already checked in ErrorIfUnsupportedAlterTableStmt.
-			 */
-			Assert(list_length(commandList) <= 1);
-
 			rightRelationId = RangeVarGetRelid(partitionCommand->name, NoLock, false);
+			char rightRelationKind = get_rel_relkind(rightRelationId);
 
 			/*
-			 * Do not generate tasks if relation is distributed and the partition
-			 * is not distributed. Because, we'll manually convert the partition into
-			 * distributed table and co-locate with its parent.
+			 * We support ALTER INDEX ATTACH PARTITION and ALTER TABLE ATTACH PARTITION 
+			 * if it is only subcommand of ALTER TABLE command. Since the right relation
+			 * type is index for ALTER INDEX ATTACH PARTITION, we need to use the relation
+			 * id this index is created for.
+			 *
+			 * Both were already checked in ErrorIfUnsupportedAlterIndexStmt and
+			 * ErrorIfUnsupportedAlterTableStmt.
 			 */
-			if (!IsCitusTable(rightRelationId))
+			if (rightRelationKind == RELKIND_INDEX)
```

Comment:
when we support multi-level partitioning, I think this can also include `RELKIND_PARTITIONED_INDEX`, but let's leave as is for now